File: Go/day1/AOC.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func part1(input []byte) int {
	lines := strings.Split(string(input), "\n")
	sum := 0
	for _, line := range lines {
		nums := regexp.MustCompile("[0-9]").FindAllString(line, -1)
		if len(nums) == 0 {
			// fmt.Println(line)
			break // this is the empty line at the end of the file
		}
		num, _ := strconv.Atoi(nums[0] + nums[len(nums)-1])
		sum += num
	}
	return sum
}

func part2(input []byte) int {
	numToInt := map[string]string{
		"one":   "o1e",
		"two":   "t2o",
		"three": "th3ee",
		"four":  "f4ur",
		"five":  "fi5e",
		"six":   "s6x",
		"seven": "se7en",
		"eight": "e8ght",
		"nine":  "n9ne",
	}
	sum := 0
	lines := strings.Split(string(input), "\n")
	for _, line := range lines {
		if line == "" {
			break // this is the empty line at the end of the file
		}
		for word, num := range numToInt {
			line = strings.ReplaceAll(line, word, num)
		}
		nums := regexp.MustCompile("[0-9]").FindAllString(line, -1)
		num, _ := strconv.Atoi(nums[0] + nums[len(nums)-1])
		sum += num
	}
	return sum
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	part1_answer := part1(input)
	fmt.Printf("Part 1: %d\n", part1_answer)

	part2_answer := part2(input)
	fmt.Printf("Part 2: %d\n", part2_answer)
}

File: Python/day9/AOC.py
import re

def parse_input(input):
    return [list(map(int, line.split(" "))) for line in input] 

def getNextNumber(input):
    if all(x == 0 for x in input):
        return 0
    diff = [input[i+1] - input[i] for i in range(len(input)-1)]
    return input[-1] + getNextNumber(diff)

def getPrevNumber(input):
    if all(x == 0 for x in input):
        return 0
    diff = [input[i+1] - input[i] for i in range(len(input)-1)]
    return input[0] - getPrevNumber(diff)

def part1(input):
    next_nums = []
    for line in input:
        x = getNextNumber(line)
        next_nums.append(x)
    return sum(next_nums)

def part2(input):
    prev_nums = []
    for line in input:
        x = getPrevNumber(line)
        prev_nums.append(x)
    return sum(prev_nums)

def main():
    input = open("input.txt", "r").read().splitlines()
    nums = parse_input(input)
    part1_sum = part1(nums)
    part2_sum = part2(nums)
    print(f"Part 1: {part1_sum}")
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()